finality-aleph/src/base_protocol/handler.rs

```diff
@@ -45,18 +102,98 @@ impl Handler {
             })
             .collect()
     }
-}
 
-#[cfg(test)]
-mod tests {
-    use crate::{aleph_primitives::Block, base_protocol::handler::Handler};
-
-    #[test]
-    fn initially_no_peers() {
-        let handler = Handler::new();
-        assert!(
-            handler.peers_info::<Block>().is_empty(),
-            "there should be no peers initially"
-        );
+    fn verify_connection(
+        &self,
+        peer_id: PeerId,
+        handshake: Vec<u8>,
+        is_inbound: bool,
+    ) -> Result<Roles, ConnectError> {
+        let handshake = BlockAnnouncesHandshake::<B>::decode_all(&mut &handshake[..])
+            .map_err(|_| ConnectError::BadlyEncodedHandshake)?;
+        if handshake.genesis_hash != self.genesis_hash {
+            return Err(ConnectError::BadHandshakeGenesis);
+        }
+
+        if self.peers.contains_key(&peer_id) {
+            return Err(ConnectError::PeerAlreadyConnected);
+        }
+
+        if !self.reserved_nodes.contains(&peer_id) {
+            // check slot constraints for full non-reserved nodes
+            if is_inbound
+                && handshake.roles.is_full()
+                && self.num_full_in_peers >= self.max_full_in_peers
+            {
+                return Err(ConnectError::TooManyFullInboundPeers);
+            }
+            if !is_inbound
+                && handshake.roles.is_full()
+                && self.num_full_out_peers >= self.max_full_out_peers
+            {
+                return Err(ConnectError::TooManyFullOutboundPeers);
+            }
+            // check slot constraints for light nodes
+            if handshake.roles.is_light() && self.num_light_peers >= self.max_light_peers {
+                return Err(ConnectError::TooManyLightPeers);
+            }
+        }
```

Comment:
I'd suggest rewriting this into a single `match`, sth like ` match (is_inbound, is_full, is_in_saturated, is_out_saturated, is_light_saturated)`, although maybe it's too much conditions - feel free to resolve if you disagree

Reply:
Hmm, this would have to cover (even if covering is `_ => {}`) cases which don't occur.
I'd rather stay with the `if` but let me actually convert the 2nd and 3rd branch to `else if` so it's emphasized that only one of them is possible.

Reply:
Uh, I actually find the `else if`s confusing – especially here, when they are paired with `return`s, below they are fine.

Reply:
ok, reverted to `if ... if ... if ...` in `verify_connection` only